### PR TITLE
Refactor Blob Storage services

### LINKED ISSUE #412: Implement feature to delete entire directories in DO Spaces

### PR DESCRIPTION
**Pull request by pnobbe**
### Merge when:

* [x] The merge build succeeded
* [ ] The PR has been reviewed and is approved
* [ ] All packages are upgraded to the latest version (only PR into master)

**What issue does this PR address?**

- Introduces new functionality to our IBlobStorageService which allows for deletion of entire directories.
- Refactors existing functionality in the `SpacesBlobStorageService.cs` to be more optimal.

Fixes #412 and #411 

**Does this PR introduce a breaking change?**

No

**Additional information**:

Remove branch after merge

### CODE REVIEW

File: src/FunderMaps.Core/Interfaces/IBlobStorageService.cs [hunk 1]
```diff
@@ -49,14 +49,21 @@ public interface IBlobStorageService
 
         // FUTURE: Refactor
         /// <summary>
-        ///     Stores a file in a Digital Ocean Space.
+        ///     Stores a directory in a Digital Ocean Space.
```
Comment (yorickdewid) [inline src/FunderMaps.Core/Interfaces/IBlobStorageService.cs:52]: Amazon S3

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs [hunk 2]
```diff
@@ -23,6 +25,9 @@ namespace FunderMaps.Infrastructure.Storage
     /// </remarks>
     internal class SpacesBlobStorageService : IBlobStorageService, IDisposable
     {
+        private static readonly byte MaxKeys = Byte.MaxValue;
```
Comment (yorickdewid) [inline src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:28]: Initialize to a constant. The fact that `MaxValue` happens to be 255 cannot be assumed.

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs [hunk 3]
```diff
@@ -189,26 +194,109 @@ public async Task StoreDirectoryAsync(string directoryName, string directoryPath
...
+                TransferUtilityUploadDirectoryRequest request = new()
+                {
+                    BucketName = _options.BlobStorageName,
+                    Directory = directoryPath,
+                    KeyPrefix = directoryName,
+                    SearchOption = SearchOption.AllDirectories,
+                    UploadFilesConcurrently = true,
+                };
+
+                request.UploadDirectoryFileRequestEvent += (sender, e) =>
```
Comment (yorickdewid) [inline src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:206]: What is `e`? Write the full var name

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs [hunk 4]
```diff
@@ -189,26 +194,109 @@ public async Task StoreDirectoryAsync(string directoryName, string directoryPath
...
+                    e.UploadRequest.CannedACL = (storageObject?.IsPublic ?? false) ? S3CannedACL.PublicRead : S3CannedACL.Private;
+                    e.UploadRequest.Headers.ContentType = storageObject?.ContentType ?? e.UploadRequest.Headers.ContentType;
+                    e.UploadRequest.Headers.CacheControl = storageObject?.CacheControl ?? e.UploadRequest.Headers.CacheControl;
+                    e.UploadRequest.Headers.ContentDisposition = storageObject?.ContentDisposition ?? e.UploadRequest.Headers.ContentDisposition;
+                    e.UploadRequest.Headers.ContentEncoding = storageObject?.ContentEncoding ?? e.UploadRequest.Headers.ContentEncoding;
+                };
+
+                TransferUtilityConfig config = new()
+                {
+                    /** Note: This is currently set to the default value of 10. I did some benchmarking on 23 dec 2020
```
Comment (yorickdewid) [inline src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:217]: We're using XDC docs and C/C++ comment. Use either of those, not javadoc. Remove name.

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs [hunk 5]
```diff
@@ -189,26 +194,109 @@ public async Task StoreDirectoryAsync(string directoryName, string directoryPath
...
+
+                    request.ContinuationToken = response.NextContinuationToken;
+                } while (response.IsTruncated);
+
+                await Task.WhenAll(tasklist.ToArray());
             }
             catch (AmazonS3Exception e)
             {
-                _logger.LogError(e, "Could not store file to Spaces using S3");
+                _logger.LogTrace(e, "Could not delete directory on Spaces using S3");
```
Comment (yorickdewid) [inline src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:297]: There is no need to print the stacktrace here. Remove the `e` from logTrace. This is true in all of the solution

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs [hunk 6]
```diff
@@ -189,26 +194,109 @@ public async Task StoreDirectoryAsync(string directoryName, string directoryPath
...
+                {
+                    BucketName = _options.BlobStorageName,
+                    Prefix = directoryPath,
+                    MaxKeys = MaxKeys
+                };
+
+                List<Task> tasklist = new();
+
+                ListObjectsV2Response response;
+                do
```
Comment (yorickdewid) [inline src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:261]: This should really be a for(..). There is an initializer `await client.ListObjectsV2Async(request);`, an advance `request.ContinuationToken = response.NextContinuationToken;` and a test `response.IsTruncated`

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs [hunk 7]
```diff
@@ -189,26 +194,109 @@ public async Task StoreDirectoryAsync(string directoryName, string directoryPath
...
+                           - Patrick
+                    **/
+                    ConcurrentServiceRequests = ConcurrentServiceRequests
+                };
+
+                await new TransferUtility(client, config).UploadDirectoryAsync(request);
+            }
+            catch (AmazonS3Exception e)
+            {
+                _logger.LogError(e, "Could not store directory to Spaces using S3");
```
Comment (yorickdewid) [inline src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:232]:
```suggestion
                _logger.LogError("Could not store directory to Spaces using S3");
```

No need for stacktrace dump. The exception is captured in the next line.

**Review by yorickdewid — CHANGES_REQUESTED** [review]

File: src/FunderMaps.Core/Interfaces/IBlobStorageService.cs (hunk 1 above)
Comment (yorickdewid) [reply src/FunderMaps.Core/Interfaces/IBlobStorageService.cs:52]: The fact that this driver will communicate to another service URL (Digital Ocean) is a config matter. There should not be any mention of Digital Ocean in the code

File: src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs (hunk 3 above)
Comment (pnobbe) [reply src/FunderMaps.Infrastructure/Storage/SpacesBlobStorageService.cs:206]: `e` is the full var name. 
image

**Review by yorickdewid — APPROVED** [review]

**Pull request merged**